store/postgres/src/relational_queries.rs

```diff
@@ -1005,10 +1077,200 @@ impl<'a> FilterQuery<'a> {
+            self.order_by(out)?;
+            self.limit(out);
+        }
+        Ok(())
+    }
+
+    fn matches_to_json(&self, out: &mut AstPass<Pg>) -> QueryResult<()> {
+        for (i, (table, _)) in self.table_filter_pairs.iter().enumerate() {
+            if i > 0 {
+                out.push_sql("\nunion all\n");
+            }
+            out.push_sql("select m.entity, to_jsonb(e.*) as data, e.id");
+            self.select_sort_key_and_window(out)?;
+            out.push_sql("\n  from ");
+            out.push_identifier(&self.schema)?;
+            out.push_sql(".");
+            out.push_identifier(table.name.as_str())?;
+            out.push_sql(" e,");
+            out.push_sql(" matches m");
+            out.push_sql("\n where e.vid = m.vid and m.entity = ");
+            out.push_bind_param::<Text, _>(&table.object)?;
+            self.limit_per_window("m.pos", "\n   and ", out);
+        }
+        out.push_sql("\n ");
+        Ok(())
+    }
+    /// Only one table/filter pair, and no window
+    ///
+    /// Generate a query
+    ///   select '..' as entity, to_jsonb(e.*) as data
+    ///     from table e
+    ///    where filter
+    ///    order by .. limit .. skip ..
+    fn query_no_window_one_entity(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        let (table, filter) = self
+            .table_filter_pairs
+            .first()
+            .expect("we already checked that there is exactly one table");
+        Self::select_entity_and_data(table, &mut out)?;
+        self.filtered_rows(table, filter, out.reborrow())?;
+        self.order_by(&mut out)?;
+        self.limit(&mut out);
+        Ok(())
+    }
+
+    /// Only one table/filter pair, and a window
+    ///
+    /// Generate a query
+    ///   select '..' as entity, to_jsonb(e.*) as data
+    ///     from (select *, rank() over (partition by window order by ..) as pos
+    ///             from table e
+    ///            where filter) e
+    ///     where e.pos > skip and e.pos <= first + skip
+    ///     order by ..
+    fn query_window_one_entity(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        let (table, filter) = self
+            .table_filter_pairs
+            .first()
+            .expect("we already checked that there is exactly one table");
+        Self::select_entity_and_data(table, &mut out)?;
+        out.push_sql(" from (\n");
+        out.push_sql("select *");
+        self.rank_window(&mut out)?;
+        self.filtered_rows(table, filter, out.reborrow())?;
+        out.push_sql(") e");
+        self.limit_per_window("e.pos", "where ", &mut out);
+        out.push_sql("\n ");
+        self.order_by(&mut out)
+    }
+
+    fn query_no_window(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        // We have multiple tables which might have different schemas since
+        // the entity_types come from implementing the same interface. We
+        // need to do the query in two steps: first we build a CTE with the
```

Comment:
What's a CTE?

Reply:
It's a [Common Table Expression](https://www.postgresql.org/docs/11/queries-with.html)